### PR TITLE
Fix MSCDEX regression in game "Fascination (1991)"

### LINKED ISSUE #3162: Possible Regression in Fascination (1991)
### Are you using the latest Dosbox-Staging Version?

- [X] I have checked releases and am using the latest release.

### Different version than latest?

0.81.0-alpha

### What Operating System are you using?

Windows 11

### If Other OS, please describe

_No response_

### Relevant hardware info

_No response_

### Have you checked that no other similar issue already exists?

- [X] I have searched and not found similar issues.

### A clear and concise description of what the bug is.

Using exodos v6 release of the game staging 80.1 plays the CD version (using exodos, select option 1 to play using DOSBox, then option 2 for CD version. At the config screen press enter) no problem. However in 81.0 alpha you get the following screenshot after the config menu closes:
image

### Steps to reproduce the behaviour.

As above

### Download URL of affected game or software

exodos v6

### Your configuration

```ini
exodos v6
```

### Provide a Log

```text
Nothing relevant
```

### Code of Conduct & Contributing Guidelines

- [X] Yes, I agree.

### PR DESCRIPTION
**Pull request by weirddan455**
# Description
Regression from commit c4a6c7db9abe3ebf84953744794b4b82266732e9

Fixed a bug when traversing the DOS device linked list. The last valid device is one before the end pointer.

~~Do not merge yet.  This does fix #3162 but re-introduces bug #2304  All this DOS memory stuff is making my head hurt so this as far as I've got tonight.  Wanted to post this in case anyone has any suggestions.  I suspect someone is clobbering the linked list somewhere.~~

Turns out DTA wasn't being allocated enough pages and was the culprit stomping on this list.  Fixed that underlying bug and everything looks good now.

## Related issues
#2304 
#2310
#3162

# Manual testing
Reproduced bug #3162 on main.  This PR fixes the bug by correctly appending to the linked list.

Reproduced #2304 one commit before the #2310 PR.  Confirmed this bug does not occur on main.  ~~This PR makes this bug re-appear~~   Root cause of this bug is now fixed in the first commit of this PR (not enough pages allocated by DTA).

# Checklist

_Please tick the items as you have addressed them. Don't remove items; leave the ones that are not applicable unchecked._

I have:

- [x] followed the project's [contributing guidelines](https://github.com/dosbox-staging/dosbox-staging/blob/master/CONTRIBUTING.md) and [code of conduct](https://github.com/dosbox-staging/dosbox-staging/blob/master/CODE_OF_CONDUCT.md).
- [x] performed a self-review of my code.
- [x] commented on the particularly hard-to-understand areas of my code.
- [x] split my work into well-defined, bisectable commits, and I [named my commits well](https://github.com/dosbox-staging/dosbox-staging/blob/main/CONTRIBUTING.md#commit-messages).
- [x] applied the appropriate labels (bug, enhancement, refactoring, documentation, etc.)
- [x] [checked](https://github.com/dosbox-staging/dosbox-staging/blob/main/scripts/compile_commits.sh) that all my commits can be built.
- [ ] confirmed that my code does not cause performance regressions (e.g., by running the Quake benchmark).
- [ ] added unit tests where applicable to prove the correctness of my code and to avoid future regressions.
- [ ] made corresponding changes to the documentation or the website according to the [documentation guidelines](https://github.com/dosbox-staging/dosbox-staging/blob/main/DOCUMENTATION.md).
- [ ] [locally verified](https://github.com/dosbox-staging/dosbox-staging/blob/main/DOCUMENTATION.md#previewing-documentation-changes-locally) my website or documentation changes.

### CODE REVIEW

**Comment by weirddan455** [comment]
I'll try to look at this more tomorrow.  I think I've fixed a bug.  Before, the append function was always writing into memory address 0xFFFFFFFF (I guess DOS uses this instead of a traditional NULL (0) pointer to denote the end of a linked list).

However, it brings back the bug with the never-ending loop elsewhere in the code.  I'll probably need to research how real DOS manages this list.  All these `read_read*`/`read_write*` and `Segment` `Offset` things are making my head hurt.  I'm not even sure how to examine this weird DOS memory in a debugger to figure out what's going on.

@johnnovak @kcgen if you have any suggestions, fire away.

**Comment by kcgen** [comment]
Brilliant work getting up to speed on that DOS device structure! I was worried we'd have to revert it from the refactored function to what it once was: https://github.com/dosbox-staging/dosbox-staging/pull/2310/commits/ae3691cad776e880f3174033b008828e00580290), so I'm very happy you found this end-pointer (versus end-device) difference.

Maybe there's some other aspect that real DOS does to stop the device loop - for example, if the memory space only has room for say, 256 devices, then just break out after a countdown. Something like that.

> All these `real_read*/real_write*` and Segment Offset things are making my head hurt.

Yup, these are DOS's internal memory management calls revealed. My understanding is these are very robust and correct; some DOS games and software can run for weeks and months (and longer) with these grinding away like clockwork.

It probably makes sense to have an assert loop that can re-validate various DOS memory structures literally every `real_write` operation and break instantly to catch bugs (especially on the DOSBox-emulating-DOS side, where new devices are created and injected into the chain).

**Comment by johnnovak** [comment]
> > Yup, these are DOS's internal memory management calls revealed. My understanding is these are very robust and correct; some DOS games and software can run for weeks and months (and longer) with these grinding away like clockwork.
> 
> Yeah, I don't think there's anything wrong with these functions. It just makes it harder to debug. If it was a normal host side linked list, I would maybe set a GDB watch point and see where it gets modified. Not sure how to do that with the DOS memory.
> 
> But it's very late here. I'll need to look at it with some fresh eyes tomorrow. Maybe I can debug with assert statements like you were suggesting.

Great progress @weirddan455 🚀

**Comment by weirddan455** [comment]
Did some debugging work on this.  I wrote a function to print out the linked list in a human readable way.  After `DOS_AppendDevice` gets called to append the MSCDEX drive, the list looks correct.

`CON -> MSCDEX -> EndPointer`

However, the memory gets stomped on later leading to the infinite loop bug in `DOS_CheckExtDevice`

After fiddling with GDB for a bit, I set a breakpoint where MSCDEX's "next" pointer gets set, stepped through the DOS memory calls to get the real host-side memory address, and set a watchpoint.

```
Thread 1 "dosbox" hit Hardware watchpoint 2: -location *arr

Old value = 255 '\377'
New value = 83 'S'
host_writeb (arr=0x7fffe82c7d60 "S\377\377\377", val=83 'S') at ../../include/mem_host.h:118
118     }
(gdb) bt
#0  host_writeb(uint8_t*, uint8_t) (arr=0x7fffe82c7d60 "S\377\377\377", val=83 'S') at ../../include/mem_host.h:118
#1  0x0000555556025b2a in mem_writeb_inline(PhysPt, uint8_t) (address=822608, val=83 'S') at ../../include/paging.h:371
#2  0x00005555560260aa in MEM_BlockWrite(unsigned int, void const*, unsigned long) (pt=822609, data=0x7fffffffa50b, size=9)
    at ../../src/hardware/memory.cpp:223
#3  0x0000555555dd5b9e in DOS_DTA::SetResult(char const*, unsigned int, unsigned short, unsigned short, FatAttributeFlags)
    (this=0x5555591a42d0, found_name=0x7fffffffa50b "INSTALL.COM", found_size=5137, found_date=7492, found_time=33469, found_attr=...) at ../../src/dos/dos_classes.cpp:361
#4  0x0000555555e2a926 in fatDrive::FindNextInternal(unsigned int, DOS_DTA&, direntry*)
    (this=0x55555928d790, dirClustNumber=0, dta=..., foundEntry=0x7fffffffa590) at ../../src/dos/drive_fat.cpp:1344
#5  0x0000555555e27b6d in fatDrive::getFileDirEntry(char const*, direntry*, unsigned int*, unsigned int*, bool)
    (this=0x55555928d790, filename=0x7fffffffa6e0 "INSTALL.COM", useEntry=0x7fffffffa680, dirClust=0x7fffffffa678, subEntry=0x7fffffffa67c, dir_ok=false) at ../../src/dos/drive_fat.cpp:488
#6  0x0000555555e29ec3 in fatDrive::FileExists(char const*) (this=0x55555928d790, name=0x7fffffffa6e0 "INSTALL.COM")
    at ../../src/dos/drive_fat.cpp:1115
#7  0x0000555555de25fc in DOS_FileExists(char const*) (name=0x7fffffffa9e0 "install.COM") at ../../src/dos/dos_files.cpp:1502
#8  0x00005555561c6cf2 in DOS_Shell::Which[abi:cxx11](std::basic_string_view<char, std::char_traits<char> >) const
    (this=0x5555591895d0, name="install") at ../../src/shell/shell_misc.cpp:520
#9  0x00005555561c6537 in DOS_Shell::ExecuteProgram(std::basic_string_view<char, std::char_traits<char> >, std::basic_string_view<char, std::char_traits<char> >) (this=0x5555591895d0, name="install", args="") at ../../src/shell/shell_misc.cpp:459
#10 0x00005555561a6170 in DOS_Shell::DoCommand(char*) (this=0x5555591895d0, line=0x7fffffffbdf7 "")
    at ../../src/shell/shell_cmds.cpp:204
#11 0x000055555619da6b in DOS_Shell::ParseLine(char*) (this=0x5555591895d0, line=0x7fffffffbdf0 "install")
    at ../../src/shell/shell.cpp:335
#12 0x000055555619dedd in DOS_Shell::RunBatchFile() (this=0x5555591895d0) at ../../src/shell/shell.cpp:379
#13 0x000055555619e4d9 in DOS_Shell::Run() (this=0x5555591895d0) at ../../src/shell/shell.cpp:442
#14 0x00005555561a1325 in SHELL_Init() () at ../../src/shell/shell.cpp:1448
#15 0x0000555555c07069 in Config::StartUp() (this=0x555558a12bb0) at ../../src/misc/setup.cpp:1531
#16 0x0000555555f65666 in sdl_main(int, char**) (argc=1, argv=0x7fffffffe4f8) at ../../src/gui/sdlmain.cpp:4961
#17 0x0000555555bd9cf6 in main(int, char**) (argc=1, argv=0x7fffffffe4f8) at ../../src/main.cpp:30
```

And this is where I'm stuck.  It seems `DOS_DTA::SetResult` is stomping on the linked list setting MSCDEX's next pointer to a garbage value.

**Comment by johnnovak** [comment]
@weirddan455 @kcgen I don't want you guys to stress over this, we can absolutely release the RC with such regressions and fix them during the RC period. Don't want anyone to stress about such regressions 😄 

But yeah, this must be fixed for the final release.

**Comment by weirddan455** [comment]
Actually, I think I've found the cause.  We're not asking for enough pages here:

https://github.com/dosbox-staging/dosbox-staging/blob/4cc1c18a964a9f7f31834043e631756329f2dc10/src/dos/drive_fat.cpp#L794

The struct must have grown at some point and this hard-coded value never got updated.  Increasing the number of pages to 3 seems to fix the problem.  I'll get the PR updated shortly.

**Comment by weirddan455** [comment]
Ready for review now.  @kcgen you might want to kick your Mac.  It's giving an out of space error again in CI.

File: src/dos/drive_fat.cpp [hunk 1]
```diff
@@ -791,7 +791,11 @@ fatDrive::fatDrive(const char *sysFilename,
 	struct partTable mbrData;
 	
 	if(imgDTASeg == 0) {
-		imgDTASeg = DOS_GetMemory(2);
+		uint16_t dta_pages = sizeof(struct sDTA) / 16;
```
Comment (kcgen) [inline src/dos/drive_fat.cpp:794]: Can it be constexpr-able?
(Maybe 16 can become a named constant, too?)

**Review by kcgen — APPROVED** [review]
All looks good @weirddan455. 
Another very nice fix.

**Review by johnnovak — APPROVED** [review]
I'm not gonna spend time to understand this in detail, so if @kcgen thinks it's fine, I'm merging. Code looks fine on the surface 👍🏻

**Comment by weirddan455** [comment]
Haha well if you want the tl;dr version @johnnovak 

Commit 1:  We didn't allocate enough memory so we wrote outside the bounds of where we were supposed to.  Fix to allocate enough memory.

Commit 2:  Manually rolling linked lists is bug prone.  When it got refactored in #2310, a bug slipped in.  I fixed the bug so behavior is as it was before.

**Review by kcgen — APPROVED** [review]

**Pull request merged**